### ISSUE TITLE
webpack+JSX + generators: Did not recognize object of type "JSXElement"

### ISSUE THREAD

**Issue by Naddiseo** [babel/babel#930]
I have my own custom JSX transformer based upon the react one, except that its output it pre-compiled into the mithril's output format: `{ "tag": "TAG", "attrs", {}, "children": [] }`

If in the same file I have a generator function I'm getting the following build error:
```
Module build failed: AssertionError: ~/projects/test/test.jsx: did not recognize object of type "JSXElement"
    at Object.getFieldNames (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/types.js:595:16)
    at visitChildren (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:203:32)
    at Visitor.PVp.visitWithoutReset (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:188:16)
    at visitChildren (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:226:21)
    at Visitor.PVp.visitWithoutReset (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:188:16)
    at NodePath.each (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path.js:96:22)
    at visitChildren (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:199:14)
    at Visitor.PVp.visitWithoutReset (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:188:16)
    at visitChildren (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:226:21)
    at Visitor.PVp.visitWithoutReset (~/projects/test/node_modules/babel-core/node_modules/ast-types/lib/path-visitor.js:188:16)
```

If I remove the generator, it works fine. I've tried stepping through with node-debug and as far as I can tell, my loader is called and properly transforms the code, so I think there may be a conflict somewhere  in the regenerator transforms. My intuition is because regenerator/ast-types uses `XJS*` whereas babel uses `JSX*` for the ast tags.

I've created a gist with my webpack conf, a test file, and my jsx loader/transformer: https://gist.github.com/Naddiseo/800aab5c780d9440ba70

**Comment by fkling** [comment]
I just created a pull request for ast-types: https://github.com/benjamn/ast-types/pull/93 . I assume that something sets a custom parser (acorn-babel ?) which doesn't produce XJS* nodes anymore.

**Comment by neVERberleRfellerER** [comment]
@Naddiseo I've had some problems with your code but I have something I have to ask.
Do you know about this?

```
/** @jsx dom */

function dom(tag, attrs, ...children) { // dom - the name after @jsx, put whatever you want
    return { tag, attrs, children };
}

console.log(JSON.stringify(
<div>
  <span>Something<span>else</span></span>
  <span>Somethign else</span>
  <span>Hahah</span>
</div>, null, 2
));
```
output
```
{
  "tag": "div",
  "attrs": null,
  "children": [
    {
      "tag": "span",
      "attrs": null,
      "children": [
        "Something",
        {
          "tag": "span",
          "attrs": null,
          "children": [
            "else"
          ]
        }
      ]
    },
    {
      "tag": "span",
      "attrs": null,
      "children": [
        "Somethign else"
      ]
    },
    {
      "tag": "span",
      "attrs": null,
      "children": [
        "Hahah"
      ]
    }
  ]
}
```
I am sorry if this is know to you and does not fir zour need and I am just wasting your time. I'll look into your gist again later.

However, from quick look:
```
{( for (item of ctrl.getList()) inner(item) )}
```
This might be one problem: you need something to replace this:
```
{ items.map(item) }
```
Array or single item should be here. You do not return anything to JSX. Maybe something like this:
```
<span>{
  (function () {
    var items = [];
    for (item of ctrl.getList()) {
     items.push(inner(item));
    }
    return items;
  })()
}</span>
```

**Comment by Naddiseo** [comment]
@fkling That PR to ast-types does fix my problem, I guess my intuition was right.

@neVERberleRfellerER, I didn't know about the /** jsx */ thing until this morning, but it doesn't apply since I wanted to pre-compile the output anyway.

I guess this can be closed once the ast-types dependency has been updated.

**Comment by sebmck** [comment]
Fixed as of 4.7.0, thanks!
